### PR TITLE
Implementation of pattern matching in the building of type driven algorithms

### LINKED ISSUE #7175: Implementation of pattern matching in the building of type driven algorithms
#### Proposed title of article
Implementation of pattern matching in the building of type driven algorithms

#### Proposed article introduction
Pattern matching is the process of determining whether a specific sequence of characters, tokens, or data occurs among the given data. It can also be used to find and replace a matching pattern in a text or code with another text or code. Every application with search capabilities makes use of pattern matching in some way.

In this article, we are going to look in details what pattern matching entails and explore the various methods in which pattern matching is used in the implementation of various concepts. This will be followed by back to back examples for the purposes of description.

#### Key takeaways 
1.  A look at scenarios for pattern matching
2.  A look at pattern matching designs
3.  A look at the implementation of  the basic toll calculations

#### Article quality
The article mainly aims at providing high quality content that will be well understood by all our learners

#### References
n/a

### PR DESCRIPTION
**Pull request by Dinahanyango**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #7175

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
Plagiarism check complete

File: content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
+### Introduction
+This article is going to explore various functionalities of the C# programming language and how it is used in the implementation of various concepts but to be more specific, the article covers the use of pattern matching.
+### Table of contents
```
Comment (mohamedgh16) [inline content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md:3]: Add a space line after each header.

File: content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+- Prerequisite
+- Pattern matching
+- Pattern designs
+- Toll calculations
+- Add in the cost of occupancy
+- Addition to the pricing
+- Conclusion
+- What next
```
Comment (mohamedgh16) [inline content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md:11]: Check this link that will help you with creating your table of contents correctly.
https://stackoverflow.com/questions/11948245/markdown-to-create-pages-and-table-of-contents

File: content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+To carry on with this article, you will be required to download Visual Studio and have it installed on your machine. You can click [here](https://www.geeksforgeeks.org/setting-environment-c-sharp/) and follow the guidelines on installation.
+
+One should also have some knowledge about C# programming language.
+
+### Pattern matching
+Today's development involves various techniques which include how one perceives data and many more factors.
+
+A programming team may lack the capability to have access to the incoming data. Every data type can be represented in its unique fashion in an object-oriented environment.
+
+Once you develop an application, it will have virtual functions, new types, and abstraction and in doing so the code that will be written will be less
```
Comment (mohamedgh16) [inline content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md:22]: Fix any punctuation errors here.

File: content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+{
+    MiniBus t when ((double)t.Riders / (double)t.Capacity) < 0.50 => 5.00m + 1.00m, //If the Minibus1 condition is less than 0.50, return 5.00m + 1.00m
+    MiniBus t when ((double)t.Riders / (double)t.Capacity) > 0.90 => 5.00m - 0.99m, //If the Minibus2 condition is greater than 0.90, return 5.00m - 0.99m
+    MiniBus => 5.00m,
+};
+```
+Any number of passengers in the delivery vehicles is unimportant to the toll authority. Furthermore, the class in which the weight of the vehicle lies in what is considered when change is needed.
+- Lorries weighing more than 4500 pounds will be charged an additional $4.35.
+- Light lorries weighing less than 2500 pounds receive a $1.85 discount.
+The following code is used to implement that rule.
```
Comment (mohamedgh16) [inline content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md:123]: Try to enhance and fix any errors here.

File: content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        
+        
+    }
+}
+```
+The code described above works as required though it's not too clear and to make it as clear as possible to a reader, one must hover through the nested if statements and the input scenarios, and in conjunction with several techniques, pattern matching does it best.
+
+As an outcome, the expressions become more difficult which as a result more unreadable and accuracy is never achieved. Instead, through the combination of operations, a tuple of values encompassing the states can be formed. Therefore, if the individual methods are brought together as one, then pattern matching is applied and solves the toll calculations.
+
+The time when the charge was collected is saved in a DateTime structure by the system that collects the charge. Below is an example description.
```
Comment (mohamedgh16) [inline content/articles/implementation-of-pattern-matching-in-the-building-of-type-driven-algorithms/index.md:236]: Fix any grammatical and punctuation errors in this section.

**Review by ghost — COMMENTED** [review]
@Dinahanyango Great article, please implement the changes so we can move on to tech review.

**Comment by Dinahanyango** [comment]
@mohamedgh16 you can have a check

**Comment by hectorkambow** [comment]
@Dinahanyango  

We will be closing this article - as I was going through it for a final review. 

Many of the sentences were confusing. 

I could not truly understand the structure of the article as I was reading it. 

There were also some minor mistake in the Prerequisite section that I was not sure how it was suppose to flow.

**Pull request closed without merging**